File: backend/src/test/kotlin/org/globe42/web/charges/ChargeControllerTest.java
package org.globe42.web.charges;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

import org.globe42.dao.ChargeDao;
import org.globe42.dao.ChargeTypeDao;
import org.globe42.dao.PersonDao;
import org.globe42.domain.Charge;
import org.globe42.domain.ChargeCategory;
import org.globe42.domain.ChargeType;
import org.globe42.domain.Person;
import org.globe42.test.Answers;
import org.globe42.test.BaseTest;
import org.globe42.web.exception.BadRequestException;
import org.globe42.web.exception.NotFoundException;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

/**
 * Unit tests for {@link ChargeController}
 * @author JB Nizet
 */
public class ChargeControllerTest extends BaseTest {

    @Mock
    private PersonDao mockPersonDao;

    @Mock
    private ChargeDao mockChargeDao;

    @Mock
    private ChargeTypeDao mockChargeTypeDao;

    @InjectMocks
    private ChargeController controller;

    @Test
    public void shouldList() {
        Long personId = 42L;
        Person person = new Person(personId);
        Charge charge = createCharge(12L);
        person.addCharge(charge);
        when(mockPersonDao.findById(personId)).thenReturn(Optional.of(person));

        List<ChargeDTO> result = controller.list(personId);

        assertThat(result).hasSize(1);
        assertThat(result.get(0).getId()).isEqualTo(charge.getId());
        assertThat(result.get(0).getType().getId()).isEqualTo(charge.getType().getId());
        assertThat(result.get(0).getMonthlyAmount()).isEqualTo(charge.getMonthlyAmount());
    }

    @Test
    public void shouldThrowIfPersonNotFound() {
        Long personId = 42L;
        when(mockPersonDao.findById(personId)).thenReturn(Optional.empty());

        assertThatExceptionOfType(NotFoundException.class).isThrownBy(() -> controller.list(personId));
    }

    @Test
    public void shouldDelete() {
        Long personId = 42L;
        Long chargeId = 12L;
        Charge charge = new Charge(chargeId);
        charge.setPerson(new Person(personId));
        when(mockChargeDao.findById(chargeId)).thenReturn(Optional.of(charge));

        controller.delete(personId, chargeId);

        verify(mockChargeDao).delete(charge);
    }

    @Test
    public void shouldAcceptDeletionIfNotFoundToBeIdempotent() {
        Long chargeId = 12L;
        when(mockChargeDao.findById(chargeId)).thenReturn(Optional.empty());

        controller.delete(42L, chargeId);

        verify(mockChargeDao, never()).delete(any());
    }

    @Test
    public void shouldRejectDeletionIfNotInCorrectPerson() {
        Long chargeId = 12L;
        Long personId = 42L;
        Charge charge = new Charge(chargeId);
        charge.setPerson(new Person(personId));

        when(mockChargeDao.findById(chargeId)).thenReturn(Optional.of(charge));

        assertThatExceptionOfType(NotFoundException.class).isThrownBy(() -> controller.delete(456L, chargeId));
    }

    @Test
    public void shouldCreate() {
        Long personId = 42L;
        Long typeId = 12L;

        Person person = new Person(personId);
        ChargeType type = createChargeType(typeId);

        when(mockPersonDao.findById(personId)).thenReturn(Optional.of(person));
        when(mockChargeTypeDao.findById(typeId)).thenReturn(Optional.of(type));
        when(mockChargeDao.save(any(Charge.class))).thenAnswer(
            Answers.<Charge>modifiedFirstArgument(charge -> charge.setId(34L)));

        ChargeCommandDTO command = new ChargeCommandDTO(typeId, BigDecimal.TEN);

        ChargeDTO result = controller.create(personId, command);

        assertThat(result.getId()).isEqualTo(34L);
        assertThat(result.getMonthlyAmount()).isEqualByComparingTo(command.getMonthlyAmount());
        assertThat(result.getType().getId()).isEqualTo(command.getTypeId());
    }

    @Test
    public void shouldThrowWhenCreatingForUnknownPerson() {
        Long personId = 42L;
        Long typeId = 12L;

        Person person = new Person(personId);
        when(mockPersonDao.findById(personId)).thenReturn(Optional.of(person));
        when(mockChargeTypeDao.findById(typeId)).thenReturn(Optional.empty());

        ChargeCommandDTO command = new ChargeCommandDTO(typeId, BigDecimal.TEN);
        assertThatExceptionOfType(BadRequestException.class).isThrownBy(() -> controller.create(personId, command));
    }

    @Test
    public void shouldThrowWhenCreatingWithTooLargeAmount() {
        Long personId = 42L;
        Long typeId = 12L;

        Person person = new Person(personId);
        ChargeType type = createChargeType(typeId);
        type.setMaxMonthlyAmount(new BigDecimal("9"));
        when(mockPersonDao.findById(personId)).thenReturn(Optional.of(person));
        when(mockChargeTypeDao.findById(typeId)).thenReturn(Optional.of(type));

        ChargeCommandDTO command = new ChargeCommandDTO(typeId, BigDecimal.TEN);
        assertThatExceptionOfType(BadRequestException.class).isThrownBy(() -> controller.create(personId, command));
    }

    @Test
    public void shouldThrowWhenCreatingForUnknownType() {
        Long personId = 42L;
        Long typeId = 12L;

        when(mockPersonDao.findById(personId)).thenReturn(Optional.empty());

        ChargeCommandDTO command = new ChargeCommandDTO(typeId, BigDecimal.TEN);
        assertThatExceptionOfType(NotFoundException.class).isThrownBy(() -> controller.create(personId, command));
    }

    static Charge createCharge(Long id) {
        Charge charge = new Charge(id);
        ChargeType chargeType = createChargeType(id * 10);
        charge.setType(chargeType);
        charge.setMonthlyAmount(new BigDecimal("123.45"));
        return charge;
    }

    static ChargeType createChargeType(Long id) {
        ChargeType chargeType = new ChargeType(id);
        chargeType.setCategory(new ChargeCategory(id * 10, "rental"));
        return chargeType;
    }
}

File: backend/src/test/kotlin/org/globe42/web/charges/ChargeControllerTest.kt
package org.globe42.web.charges

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.globe42.dao.ChargeDao
import org.globe42.dao.ChargeTypeDao
import org.globe42.dao.PersonDao
import org.globe42.domain.Charge
import org.globe42.domain.ChargeCategory
import org.globe42.domain.ChargeType
import org.globe42.domain.Person
import org.globe42.web.exception.BadRequestException
import org.globe42.web.exception.NotFoundException
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.util.*

/**
 * Unit tests for [ChargeController]
 * @author JB Nizet
 */
class ChargeControllerTest {

    private val mockPersonDao = mockk<PersonDao>()

    private val mockChargeDao = mockk<ChargeDao>(relaxUnitFun = true)

    private val mockChargeTypeDao = mockk<ChargeTypeDao>()

    private val controller = ChargeController(mockPersonDao, mockChargeDao, mockChargeTypeDao)

    @Test
    fun `should list`() {
        val personId = 42L
        val person = Person(personId)
        val charge = createCharge(12L)
        person.addCharge(charge)
        every { mockPersonDao.findById(personId) } returns Optional.of(person)

        val result = controller.list(personId)

        assertThat(result).hasSize(1)
        assertThat(result[0].id).isEqualTo(charge.id)
        assertThat(result[0].type.id).isEqualTo(charge.type.id)
        assertThat(result[0].monthlyAmount).isEqualTo(charge.monthlyAmount)
    }

    @Test
    fun `should throw if person not found`() {
        val personId = 42L
        every { mockPersonDao.findById(personId) } returns Optional.empty()

        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.list(personId) }
    }

    @Test
    fun `should delete`() {
        val personId = 42L
        val chargeId = 12L
        val charge = Charge(chargeId)
        charge.person = Person(personId)
        every { mockChargeDao.findById(chargeId) } returns Optional.of(charge)

        controller.delete(personId, chargeId)

        verify { mockChargeDao.delete(charge) }
    }

    @Test
    fun `should accept deletion if not found to be idempotent`() {
        val chargeId = 12L
        every { mockChargeDao.findById(chargeId) } returns Optional.empty()

        controller.delete(42L, chargeId)

        verify(inverse = true) { mockChargeDao.delete(any()) }
    }

    @Test
    fun `should reject deletion if not in correct person`() {
        val chargeId = 12L
        val personId = 42L
        val charge = Charge(chargeId)
        charge.person = Person(personId)

        every { mockChargeDao.findById(chargeId) } returns Optional.of(charge)

        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.delete(456L, chargeId) }
    }

    @Test
    fun `should create`() {
        val personId = 42L
        val typeId = 12L

        val person = Person(personId)
        val type = createChargeType(typeId)

        every { mockPersonDao.findById(personId) } returns Optional.of(person)
        every { mockChargeTypeDao.findById(typeId) } returns Optional.of(type)
        every { mockChargeDao.save(any<Charge>()) } answers { arg<Charge>(0).apply { id = 34L } }

        val command = ChargeCommandDTO(typeId, BigDecimal.TEN)

        val result = controller.create(personId, command)

        assertThat(result.id).isEqualTo(34L)
        assertThat(result.monthlyAmount).isEqualByComparingTo(command.monthlyAmount)
        assertThat(result.type.id).isEqualTo(command.typeId)
    }

    @Test
    fun `should throw when creating for unknown person`() {
        val personId = 42L
        val typeId = 12L

        val person = Person(personId)
        every { mockPersonDao.findById(personId) } returns Optional.of(person)
        every { mockChargeTypeDao.findById(typeId) } returns Optional.empty()

        val command = ChargeCommandDTO(typeId, BigDecimal.TEN)
        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.create(personId, command) }
    }

    @Test
    fun `should throw when creating with too large amount`() {
        val personId = 42L
        val typeId = 12L

        val person = Person(personId)
        val type = createChargeType(typeId)
        type.maxMonthlyAmount = BigDecimal("9")
        every { mockPersonDao.findById(personId) } returns Optional.of(person)
        every { mockChargeTypeDao.findById(typeId) } returns Optional.of(type)

        val command = ChargeCommandDTO(typeId, BigDecimal.TEN)
        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.create(personId, command) }
    }

    @Test
    fun `should throw when creating for unknown type`() {
        val personId = 42L
        val typeId = 12L

        every { mockPersonDao.findById(personId) } returns Optional.empty()

        val command = ChargeCommandDTO(typeId, BigDecimal.TEN)
        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.create(personId, command) }
    }
}

internal fun createCharge(id: Long): Charge {
    val charge = Charge(id)
    val chargeType = createChargeType(id * 10)
    charge.type = chargeType
    charge.monthlyAmount = BigDecimal("123.45")
    return charge
}

internal fun createChargeType(id: Long): ChargeType {
    val chargeType = ChargeType(id)
    chargeType.category = ChargeCategory(id * 10, "rental")
    chargeType.name = "type $id"
    return chargeType
}
